python/runtime/optimize/optimize.py

```diff
@@ -469,3 +530,243 @@ def run_optimize(datasource, select, variables, variable_type,
                              result_value_name=result_value_name,
                              datasource=datasource,
                              result_table=result_table)
+
+
+OPTFLOW_HTTP_HEADERS = {
+    'content-type': 'application/json',
+    'accept': 'application/json',
+}
+
+
+def query_optflow_job_status(url, record_id, user_number, token):
+    url = "{}?userNumber={}&recordId={}&token={}".format(
+        url, user_number, record_id, token)
+    response = requests.get(url, headers=OPTFLOW_HTTP_HEADERS)
+    response.raise_for_status()
+    response_json = response.json()
+    if not response_json['success']:
+        raise ValueError('cannot get status of job {}'.format(record_id))
+
+    return response_json['data']['status'].lower()
+
+
+def query_optflow_job_log(url, record_id, user_number, token, start_line_num):
+    url = "{}?userNumber={}&recordId={}&token={}".format(
+        url, user_number, record_id, token)
+    response = requests.get(url, headers=OPTFLOW_HTTP_HEADERS, stream=True)
+    response.raise_for_status()
+    response_json = response.json()
+    if not response_json['success']:
+        raise ValueError('cannot get log of job {}'.format(record_id))
+
+    logs = response_json['data']['logs']
+    end_line_num = len(logs)
+
+    # NOTE(sneaxiy): ascii(log) is necessary because the character inside
+    # log may be out of the range of ASCII characters.
+    # The slice [1:-1] is used to remove the quotes. e.g.:
+    # original string "abc" -> ascii("abc") outputs "'abc'"
+    # -> the slice [1:-1] outputs "abc"
+    logs = [ascii(log)[1:-1] for log in logs[start_line_num:]]
+    return logs, end_line_num
+
+
+def print_job_log_till_finish(status_url, log_url, record_id, user_number,
+                              token):
+    def call_func_with_retry(func, times):
+        for _ in six.moves.range(times - 1):
+            try:
+                return func()
+            except:
+                pass
+
+        return func()
+
+    status = None
+    line_num = 0
+    while True:
+        query_status = lambda: query_optflow_job_status(
+            status_url, record_id, user_number, token)
+        query_log = lambda: query_optflow_job_log(log_url, record_id,
+                                                  user_number, token, line_num)
+        status = call_func_with_retry(query_status, 3)
+        logs, line_num = call_func_with_retry(query_log, 3)
+
+        for log in logs:
+            print(log)
+
+        # status may be 'success', 'failed', 'running', 'prepare'
+        if status in ['success', 'failed']:
+            break
+
+        time.sleep(2)  # sleep for some times
+
+    return status == 'success'
+
+
+def send_optflow_http_request(train_table, result_table, fsl_file_content,
```

Comment:
Put these functions in another file?

Reply:
Done. Removed to be `python/runtime/optimize/optflow_submit.py`.